tower-balance/src/load/peak_ewma.rs

```diff
@@ -0,0 +1,373 @@
+use futures::{Async, Poll};
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+use tokio::clock;
+use tower_discover::{Change, Discover};
+use tower_service::Service;
+
+use super::{Instrument, InstrumentFuture, NoInstrument};
+
+use Load;
+
+/// Wraps an `S`-typed Service with Peak-EWMA load measurement.
+///
+/// `PeakEwma` implements `Load` with the `Cost` metric that estimates the amount of
+/// pending work to an endpoint. Work is calculated by multiplying the
+/// exponentially-weighted moving average (EWMA) of response latencies by the number of
+/// pending requests. The Peak-EWMA algorithm is designed ot be especially sensitive to
+/// worst-case latencies. Over time, the peak latency value decays towards the moving
+/// average of latencies to the endpoint.
+///
+/// As requests are sent to the underlying service, an `I`-typed instrumentation strategy
+/// is used to track responses to measure latency in an application-specific way. The
+/// default strategy measures latency as the elapsed time from the request being issued to
+/// the underlying service to the response future being satisfied (or dropped).
+///
+/// When no latency information has been measured for an endpoint, the [standard RTT
+/// within a datacenter][numbers] (500ns) is used.
+///
+/// [numbers]: https://people.eecs.berkeley.edu/~rcs/research/interactive_latency.html
+pub struct PeakEwma<S, I = NoInstrument> {
+    service: S,
+    decay_ns: f64,
+    rtt_estimate: Arc<Mutex<Option<RttEstimate>>>,
+    instrument: I,
+}
+
+/// Wraps a `D`-typed stream of discovery updates with `PeakEwma`.
+pub struct WithPeakEwma<D, I = NoInstrument> {
+    discover: D,
+    decay_ns: f64,
+    instrument: I,
+}
+
+/// Represents the relative cost of communicating with a service.
+///
+/// The underlying value estimates the amount of pending work to a service: the Peak-EWMA
+/// latency estimate multiplied by the number of pending requests.
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd)]
+pub struct Cost(f64);
+
+/// Updates `RttEstimate` when dropped.
+pub struct Handle {
+    sent_at: Instant,
+    decay_ns: f64,
+    rtt_estimate: Arc<Mutex<Option<RttEstimate>>>,
+}
+
+/// Holds the current RTT estimateand the last time this value was updated.
+struct RttEstimate {
+    update_at: Instant,
+    rtt_ns: f64,
+}
+
+/// The "standard" RTT within a datacenter is ~500ns.
+///
+/// https://people.eecs.berkeley.edu/~rcs/research/interactive_latency.html
+const DEFAULT_RTT_ESTIMATE: f64 = 500_000.0;
+
+const NANOS_PER_MILLI: f64 = 1_000_000.0;
+
+// ===== impl PeakEwma =====
+
+impl<D: Discover> WithPeakEwma<D, NoInstrument> {
+    pub fn new(discover: D, decay: Duration) -> Self {
```

Comment:
Is it going to be common to create an instance with NoInstrument?

Reply:
Probably not enough to warrant the split